tests/tests_fabric/strategies/test_single_device.py

```diff
@@ -52,3 +55,69 @@ def test_single_device_module_to_device():
     module = Mock(spec=torch.nn.Module)
     strategy.module_to_device(module)
     module.to.assert_called_with(strategy.root_device)
+
+
+class _MyFabricGradNorm(BoringFabric):
+    def after_backward(self, model: _FabricModule, optimizer: _FabricOptimizer):
+        self.clip_gradients(model, optimizer, max_norm=0.05, error_if_nonfinite=True)
+
+        parameters = model.parameters()
+        grad_norm = torch.linalg.vector_norm(
+            torch.stack([torch.linalg.vector_norm(p.grad.detach(), 2, dtype=torch.float32) for p in parameters]),
+            2,
+        )
+        torch.testing.assert_close(grad_norm, torch.tensor(0.05, device=self.device))
+
+    def run(self):
+        while True:
+            try:
+                super().run()
+                break
+            except RuntimeError:  # nonfinite grads -> skip and continue
+                pass
+
+
+class _MyFabricGradVal(BoringFabric):
+    def after_backward(self, model, optimizer):
+        for p in model.parameters():
+            if p.grad is not None and torch.isnan(p.grad).any().item() or torch.isinf(p.grad).any().item():
+                raise RuntimeError("Nonfinite grads")
+
+        self.clip_gradients(model, optimizer, clip_val=1e-10)
+
+        parameters = model.parameters()
+        grad_max_list = [torch.max(p.grad.detach().abs()) for p in parameters]
+        grad_max = torch.max(torch.stack(grad_max_list))
+        torch.testing.assert_close(grad_max.abs(), torch.tensor(1e-10, device=self.device))
+
+    def run(self):
+        while True:
+            try:
+                super().run()
+                break
+            except RuntimeError:  # nonfinite grads -> skip and continue
+                pass
```

Comment:
Can we use a different pattern? As I understand, you are trying to assert that nonfinite grads is not raised?

But if it is, then this will run endlessly?

Reply:
The issue here is, that with calling backward in an fp16 setting, the scaling-factor may be too large so that it produces an overflow.

In following steps the scaler will then automatically reduce this factor until no more overflow occurs so it shouldn't run endlessly.

Reply:
```suggestion
            except RuntimeError:
                # nonfinite grads -> skip and continue
                # this may repeat until the sacler finds a factor where overflow is avoided, so the while loop should eventually break
                pass
```